miden-lib/src/notes/mod.rs

```diff
@@ -1,99 +1,90 @@
 
-    let (note_script, _) = NoteScript::new(note_script_ast, &note_assembler)?;
-
-    Note::new(note_script.clone(), &inputs, &assets, serial_num, sender, tag.unwrap_or(ZERO))
+    Note::new(note_script, &inputs, &assets, serial_num, sender, tag)
 }
 
-/// Utility function generating RECIPIENT for the P2ID note script created by the SWAP script
-fn build_p2id_recipient(target: AccountId, serial_num: Word) -> Result<Digest, NoteError> {
-    // TODO: add lazy_static initialization or compile-time optimization instead of re-generating
-    // the script hash every time we call the SWAP script
-    let assembler = TransactionKernel::assembler();
-
-    let p2id_bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/note_scripts/P2ID.masb"));
-
-    let note_script_ast =
-        ProgramAst::from_bytes(p2id_bytes).map_err(NoteError::NoteDeserializationError)?;
-
-    let (note_script, _) = NoteScript::new(note_script_ast, &assembler)?;
-
-    let script_hash = note_script.hash();
+/// Generates a P2IDR note - pay to id with recall after a certain block height.
+/// This script enables the transfer of assets from one account `sender` to another account `target`
+/// additionally it adds the possibility of a recall window enabling reclaiming of assets if the
+/// note has not been consumed by the `target` in the inputed timeframe
+pub fn create_p2idr_note<R: FeltRng>(
+    sender: AccountId,
+    target: AccountId,
+    assets: Vec<Asset>,
+    recall_height: u32,
+    mut rng: R,
+) -> Result<Note, NoteError> {
+    let bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/note_scripts/P2IDR.masb"));
+    let note_script = build_note_script(bytes)?;
 
-    let serial_num_hash = Hasher::merge(&[serial_num.into(), Digest::default()]);
+    let inputs = [target.into(), recall_height.into(), ZERO, ZERO];
+    let tag: Felt = target.into();
+    let serial_num = rng.draw_word();
 
-    let merge_script = Hasher::merge(&[serial_num_hash, script_hash]);
+    Note::new(note_script.clone(), &inputs, &assets, serial_num, sender, tag)
+}
 
-    Ok(Hasher::merge(&[
-        merge_script,
-        Hasher::hash_elements(&[target.into(), ZERO, ZERO, ZERO]),
-    ]))
+/// Generates a SWAP note - swap of assets between two accounts.
+/// This script enables a swap of 2 assets between one account `sender` and any other account that
+/// is willing to consume the note. The consumer will receive the `offered_asset` and will create a
+/// new P2ID note with `sender` as target, containing the `requested_asset`
+pub fn create_swap_note<R: FeltRng>(
+    sender: AccountId,
+    offered_asset: Asset,
+    requested_asset: Asset,
+    mut rng: R,
+) -> Result<(Note, Word), NoteError> {
+    let bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/note_scripts/SWAP.masb"));
+    let note_script = build_note_script(bytes)?;
+
+    let repay_serial_num = rng.draw_word();
+    let recipient = utils::build_p2id_recipient(sender, repay_serial_num)?;
```

Comment:
minor nit: I'd name this `repay_recipient`. Also, I wonder if `payback` is better than `repay`.

Reply:
I believe that `payback` would be better